File: gfl.py
#!/usr/bin/python
# -*- coding: utf-8 -*

import argparse
import os
import time
import sys
import stat

VERSION = '1.0'

def init():
    parser = argparse.ArgumentParser(
        description = 'Find and print file list by specified path(s).',)

    parser.add_argument(
        '-t', '--type',
        choices = ['reg', 'dir', 'sym'],
        action = 'append',
        help = 'Specify the file type of the searching.(default: reg)')
    parser.add_argument(
        '-d', '--depth',
        type = int,
        default = -1,
        help = 'Specify the depth of the searching.(default: %(default)s)')
    parser.add_argument(
        '-r', '--reverse',
        action = 'store_true',
        default = False,
        help = 'Reverse the result.(default: %(default)s)')
    parser.add_argument(
        '-s', '--sort-by',
        choices = ['path', 'file', 'size', 'date', 'null'],
        default = 'path',
        help = 'Specify the key of sort.(default: %(default)s)')
    parser.add_argument(
        '-o', '--output',
        choices = ['abs', 'rel', 'bsn'],
        default = 'rel',
        help = 'Specify the style of output.(default: %(default)s)')
    parser.add_argument(
        '-g', '--group',
        action = 'store_true',
        default = False,
        help = 'Frouping the result.(default: %(default)s)')
    parser.add_argument(
        '--debug',
        type = int,
        default = 0,
        metavar = 'DEBUG-LEVEL',
        help = 'Set the debug level.(default: %(default)s)')
    parser.add_argument(
        '-v', '--version',
        action = 'version',
        version = '%(prog)s ' + VERSION,
        default = argparse.SUPPRESS,
        help = 'Display the version of this program.')

    parser.add_argument(
        'paths',
        nargs = '*',
        default = [os.getcwd() + os.path.sep],
        help = 'The path list.(default: current dir)')

    args = parser.parse_args()
    global FTYPE, SRCHDEP, REV, SORTBY, OUTPUT, GROUP, DEBUG
    FTYPE = args.type if args.type else ['reg']
    SRCHDEP = args.depth
    REV = args.reverse
    SORTBY = args.sort_by
    OUTPUT = args.output
    GROUP = args.group
    DEBUG = args.debug

    return [p if p[-1] == os.path.sep else p + os.path.sep for p in args.paths]

def debug(level, *msg):
    if level <= DEBUG:
        ori_time = time.time();
        crt_time = (time.strftime('%H:%M:%S',time.localtime(ori_time))
                 + '.' + str(ori_time - int(ori_time))[2:5])
        pmsg = ''
        for m in msg:
            pmsg += str(m)
        print "[DebugMessage %s %d] %s" % (crt_time, level, pmsg)

def getFileList(path, depth, ftype):
    dirlist = []

    if depth:
        if os.path.exists(path):
            for name in os.listdir(path):
                fullname = os.path.join(path, name)
                st = os.lstat(fullname)
                isappend = False

                if stat.S_ISDIR(st.st_mode):
                    dirlist += getFileList(fullname, depth - 1, ftype)
                    if 'dir' in ftype:
                        isappend = True
                elif 'reg' in ftype and stat.S_ISREG(st.st_mode):
                    isappend = True
                elif 'sym' in ftype and stat.S_ISLNK(st.st_mode):
                    isappend = True

                if isappend: dirlist.append([fullname, int(st.st_size), int(st.st_mtime)])
        else:
            sys.stderr.write('[ERROR] Can not found path: {}\n'.format(path))

    return dirlist

if __name__ == '__main__':
    dirpaths = init()
    debug(2, 'FTYPE = ', FTYPE)
    debug(2, 'SRCHDEP = ', SRCHDEP)
    debug(2, 'REV = ', REV)
    debug(2, 'SORTBY = ', SORTBY)
    debug(2, 'OUTPUT = ', OUTPUT)
    debug(2, 'GROUP = ', GROUP)
    debug(2, 'DEBUG = ', DEBUG)
    debug(2, 'dirpath = ', dirpaths)

    filelist = []
    for dirpath in sorted(dirpaths):
        debug(1, dirpath)
        ori_list = getFileList(dirpath, SRCHDEP, FTYPE)
        debug(1, len(ori_list), ' files.')
        filelist.append([dirpath, ori_list])

    frist_g = True
    for listgroup in filelist:
        debug(1, listgroup[0])

        if GROUP:
            if not frist_g: print
            print listgroup[0]
            frist_g  = False

        sorted_files = []

        if SORTBY != 'null':
            if SORTBY == 'path':
                sorted_files = sorted(listgroup[1], key = lambda x: x[0])
            elif SORTBY == 'file':
                sorted_files = sorted(listgroup[1], key = lambda x: os.path.basename(x[0]))
            elif SORTBY == 'size':
                sorted_files = sorted(listgroup[1], key = lambda x: x[1])
            elif SORTBY == 'date':
                sorted_files = sorted(listgroup[1], key = lambda x: x[2])
        else:
            sorted_files = listgroup[1]

        if REV: sorted_files.reverse()

        for sfile in sorted_files:
            debug(3, sfile)
            if OUTPUT == 'abs':
                print sfile[0]
            elif OUTPUT == 'rel':
                print sfile[0][len(listgroup[0]):]
            elif OUTPUT == 'bsn':
                print os.path.basename(sfile[0])

File: gfl.pl
#!/usr/bin/perl

use warnings;
use strict;
use Getopt::Long;
use Pod::Usage;
use 5.010;
use Cwd;
use File::Spec;
use Time::localtime;
use Time::HiRes qw(time);

my $HELP = 0;
my $FTYPE = ['reg'];
my $SRCHDEP = -1;
my $REV = 0;
my $SORTBY = 'path';
my $OUTPUT = 'rel';
my $GROUP = 0;
my $DEBUG = 0;

my @paths = init();

my %file_list;
for my $path (@paths) {
    chomp $path;
    push @{$file_list{$path}}, list_dir($path, $SRCHDEP, @$FTYPE);
}

my @sorted_list;
if ($SORTBY ne 'null') {
    if ($SORTBY eq 'path') {
        push @sorted_list, 
             [$_, [sort {
                       File::Spec->catfile($a->{$SORTBY}, $a->{file})
                       cmp
                       File::Spec->catfile($b->{$SORTBY}, $b->{file})
                   } @{$file_list{$_}}]]
            for (@paths);
    }
    elsif ($SORTBY eq 'file') {
        push @sorted_list, [$_, [sort {$a->{file} cmp $b->{file}} @{$file_list{$_}}]] for (@paths);
    }
    elsif ($SORTBY eq 'size' || $SORTBY eq 'date') {
        push @sorted_list, [$_, [sort {$a->{$SORTBY} <=> $b->{$SORTBY}} @{$file_list{$_}}]] for (@paths);
    }
}
else {
    push @sorted_list, [$_, $file_list{$_}] for (@paths);
}

my $frist_g = 1;
for my $path (@sorted_list) {
    if ($GROUP) {
        print "\n" unless $frist_g;
        print @$path[0]."\n";
        $frist_g = 0;
    }

    my $fmt_path = File::Spec->catdir(File::Spec->splitdir(@$path[0]));
    debug(2, "\$fmt_path = $fmt_path");

    @{@$path[1]} = reverse @{@$path[1]} if ($REV);

    for my $file (@{@$path[1]}) {
        if ($OUTPUT eq 'abs') {
            print File::Spec->catfile($file->{path}, $file->{file})."\n";
        }
        elsif ($OUTPUT eq 'rel') {
            print substr(
                      File::Spec->catfile($file->{path}, $file->{file}),
                      length($fmt_path) + 1,
                  )."\n";
        }
        elsif ($OUTPUT eq 'bsn') {
            print $file->{file}."\n";
        }
    }
}

sub init {
    GetOptions(
        'h|help' => \$HELP,
        't|type=s@' => \$FTYPE,
        'd|depth=i' => \$SRCHDEP,
        'r|reverse' => \$REV,
        's|sort-by=s' => \$SORTBY,
        'o|output=s' => \$OUTPUT,
        'g|group' => \$GROUP,
        'debug=i' => \$DEBUG,
    ) or pod2usage(2);

    my @paths = @ARGV ? @ARGV : (cwd());

    debug(2, "\$HELP = $HELP");

    for my $type (@$FTYPE) {
        check_option_value('-t/--type', $type, qw[reg dir sym]);
    }
    debug(2, "\$FTYPE = @$FTYPE");

    debug(2, "\$SRCHDEP = $SRCHDEP");

    debug(2, "\$REV = $REV");

    check_option_value('-s/--sort-by', $SORTBY, qw[path file size date null]);
    debug(2, "\$SORTBY = $SORTBY");

    check_option_value('-o/--output', $OUTPUT, qw[abs rel bsn]);
    debug(2, "\$OUTPUT = $OUTPUT");

    debug(2, "\$GROUP = $GROUP");
    debug(2, "\$DEBUG = $DEBUG");
    debug(2, "\＠paths = @paths");

    pod2usage(1) if $HELP;

    return @paths;
}

sub check_option_value {
    my $option = shift;
    my $value = shift;
    my @values = @_;

    unless (grep $value eq $_, @values) {
        my $optstr = join(', ', map "'$_'", @values);
        print STDERR "Value \"$value\" invalid for option $option (choose $optstr)\n";
        pod2usage(2);
    }
}

sub debug {
    my $level = shift;
    my $msg = shift;

    if ($level <= $DEBUG) {
        my $msecond = time;
        $msecond =~ s/.*\.(\d{3}).*/$1/;
        my $time_str = sprintf('%02d:%02d:%02d.%03d',
                               localtime->hour, localtime->min, localtime->sec, $msecond);
        say STDERR "[DebugMessage $time_str $level] $msg";
    }
}

sub list_dir {
    my $path = shift;
    my $depth = shift;
    my @ftype = @_;

    my @ret_list;

    if ($depth) {
        if (-d $path) {
            debug(1, "\$path = $path");
            local *DH;
            if (!opendir(DH, $path)) {
                print STDERR "Cannot opendir $path: $!\n";
                return;
            }

            for (readdir(DH)) {
                next if ($_ eq '.' || $_ eq '..');

                my $file = File::Spec->catfile($path, $_);

                if (-d $file) {
                    debug(4, 'Enter path: '.$file);
                    push @ret_list, list_dir($file, $depth - 1, @ftype);
                    next unless (grep $_ eq 'dir', @ftype);
                }
                else {
                    debug(4, "Found file: $file");

                    next unless (grep($_ eq 'reg', @ftype) && -f $file && !-l $file
                        or grep($_ eq 'sym', @ftype) && -l $file);
                }

                debug(1, "$file");
                my ($mode, $size, $mtime) = (lstat($file))[2, 7, 9];
                debug(2, "Size: $size");
                debug(2, "Mtime: $mtime");
                push @ret_list, {
                    path => $path,
                    file => $_,
                    size => $size,
                    date => $mtime,
                };
            }

            close(DH);
        }
        else {
            print STDERR "[ERROR] Can not found path: $path\n"
        }
    }
    else {
        debug(1, "Out of depth on path: $path");
    }

    return @ret_list;
}

__END__

=head1 NAME

gfl.pl - Find and print file list by specified path(s).

=head1 SYNOPSIS

gfl.pl [options] [path ...]

  Options:
  -h, --help
  -t {reg,dir,sym}, --type {reg,dir,sym}
  -d DEPTH, --depth DEPTH
  -r, --reverse
  -s {path,file,size,date}, --sort-by {path,file,size,date,null}
  -p {abs,rel,bsn}, --output {abs,rel,bsn}
  -g, --group
  --debug DEBUG-LEVEL
  -v, --version

=head1 OPTIONS

=over 8

=item B<-h, --help>

Show this help message and exit.

=item B<-t {reg, dir, sym}, --type {reg, dir, sym}>

Specify the file type of the searching.

=over 8

=item {reg}

Regular file.

=item {dir}

Directory.

=item {sym}

Symbolic link.

=back

=item B<-d DEPTH, --depth DEPTH>

Specify the depth of the searching.

=item B<-r, --reverse>

Reverse the result.

=item B<-s {path,file,size,date}, --sort-by {path,file,size,date,null}>

Specify the key of sort.

=over 8

=item {path}

Sort by path name.

=item {file}

Sort by file name.

=item {size}

Sort by size of file.

=item {date}

Sort by time of last modification.

=item {null}

Use default order.

=back

=item B<-p {abs,rel,bsn}, --optput {abs,rel,bsn}>

Specify the key of sort.

=over 8

=item {abs}

Absolute path.

=item {rel}

Relative path.

=item {bsn}

File name only.

=back

=item B<-g, --group>

Frouping the result.

=item B<--debug DEBUG-LEVEL>

Set the debug level.

=item B<-v, --version>

Display the version of this program and exit.

=back

=cut
